src/darwin/Framework/CHIP/MTRDevice.h

```diff
@@ -323,6 +323,25 @@ MTR_AVAILABLE(ios(16.1), macos(13.0), watchos(9.1), tvos(16.1))
                     queue:(dispatch_queue_t)queue
                completion:(void (^)(NSURL * _Nullable url, NSError * _Nullable error))completion
     MTR_AVAILABLE(ios(17.6), macos(14.6), watchos(10.6), tvos(17.6));
+
+/**
+ * Wait until a set of attributes reaches certain values or the provided timeout
+ * (if any) is reached.  The completion is called at that point, with an error
+ * if a timeout occurred.
+ *
+ * The completion is also called with an error on input errors.
+ *
+ * The values to wait for are represented as an array of response-values which
+ * have an MTRAttributePathKey and MTRDataKey.
+ *
+ * The timeout argument can be nil to indicate no timeout. Otherwise it must
```

Comment:
Is it a good idea to allow no timeout here, or even arbitrarily large timeouts? What if the attributes never reach the expected state?

Reply:
Then you never get notified.

This seems useful to me for the "I want to know if this ever happens" use case.  You might not be expecting it to happen, but would like to know if it does.

Reply:
Well, I'm wondering about forgotten Waiter objects hanging around in memory forever. A fire-and-forget API like this seems fine if the work is reasonably time-bounded (like I'm sending some command and then expect some state to be reached in e.g. 30 seconds). If these are going to hang around for a long time, I think we'd need to either have a way to cancel them, or expose the actual Waiter object and make the client hold the reference to it as long as it's wanted?

Reply:
Talked to some plausible API consumers and added a 5-minute cap.